x-pack/elastic-agent/pkg/agent/storage/sync_on_save_store.go

```diff
@@ -0,0 +1,81 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package storage
+
+import (
+	"io"
+	"os"
+	"runtime"
+
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/errors"
+)
+
+// SyncOnSaveStore syncs paths after successful write.
+// we use this sync store to avoid issues seen on
+// low spec windows environment where agent is faster
+// than filesystem and what we read is different(stale)
+// than what we just wrote.
+type SyncOnSaveStore struct {
+	enabled  bool
+	syncPath string
+	wrapped  Store
+}
+
+// NewSyncStore creates always syncing store.
+func NewSyncStore(wrappedStore Store, syncPath string) *SyncOnSaveStore {
+	return &SyncOnSaveStore{
+		enabled:  true,
+		syncPath: syncPath,
+		wrapped:  wrappedStore,
+	}
+}
+
+// NewWindowsSyncOnSaveStore creates windows syncing store.
+func NewWindowsSyncOnSaveStore(wrappedStore Store, syncPath string) *SyncOnSaveStore {
+	// TODO: think of windows 7 only syncing store as this is the slowest
+	return &SyncOnSaveStore{
+		enabled:  runtime.GOOS == "windows",
+		syncPath: syncPath,
+		wrapped:  wrappedStore,
+	}
+}
+
+// Save accepts a persistedConfig and saved it to a target file, to do so we will
+// make a temporary files if the write is successful we are replacing the target file with the
+// original content.
+func (s *SyncOnSaveStore) Save(in io.Reader) error {
+	if err := s.wrapped.Save(in); err != nil {
+		return err
+	}
+
+	if !s.enabled {
+		return nil
+	}
+
+	f, err := os.OpenFile(s.syncPath, os.O_RDWR, 0777)
```

Comment:
If you want to be more strict add `O_DIRECT`.

Reply:
i like that

Reply:
looks like not available everywhere so i'll skip direct